### ISSUE TITLE
Pop!_OS overrides harddisk APM on boot

### ISSUE THREAD

**Issue by Dekamir** [pop-os/pop#1807]
**Distribution:** 
Pop!_OS (version independent, tested on multiple versions)

**Related Application and/or Package Version:** 
Unknown. Possible culprits: _system76-power_ or _Intel storage driver_

**Issue Description:** 
When booted, Pop!_OS changes harddisk firmware configuration about Advanced Power Management (enables and sets a value), which causes problems with some configurations and make the OS unusable.

**Steps to reproduce:** 
Boot Pop!_OS (either live or installed).

**Expected behavior:** 
Pop!_OS should leave the settings as-is.

**Other Notes:**
This setting causes some disks to power-cycle frequently, to a point where the disk is unusable (the power-cycle happens every second possible).

- Tested on multiple disks, setups and various hardware.
- I have seen no other operating system or distribution to do this, to date.
- The changes are also permanent, meaning one has to boot to another operating system and manually disable the feature.
- This problem can be reproduced on Windows by installing Intel Rapid Storage driver software. To fix this, one needs to remove the driver software and manually disable APM via a third party program (like HDDScan).
- The disks tested are in good shape, and the computers tested have no problems.

**Comment by bflanagin** [comment]
Thanks for the report @Dekamir, We're going to need more information to track down the issue. 
1. Motherboard / CPU make and model
2. Harddrive/s make and model

Have you tried to disable system76-power and/or the intel storage driver? It may take some time to replicate your setup so any tests on your end will

**Comment by Dekamir** [comment]
**Motherboard:** MSI Z270-A PRO
**CPU:** Intel i5-7400
**Harddrives:**
Samsung HD103SJ (S246J90B425607)
Samsung HD322HJ (S1KQJ9BS900363)

I am fairly familiar with my way around Linux, but sadly not familiar enough to disable drivers.
If there are guides you could forward me to (or if it's just two terminal commands/text edits), I am tech savvy enough to get it.

**Comment by bflanagin** [comment]
No problem, you'll be using systemctl as root to disable and then re-enable system76-power. 
To disable:
`sudo systemctl disable system76-power`
To enable:
`sudo systemctl enable system76-power` 

I'd suggest rebooting after either command so that you're sure its not running when testing things. You can make sure the service is disabled/enabled by using `systemctl status system76-power` you should see a line that says something like "active ..." or "inactive ..." depending on the state of the service.

**Comment by bflanagin** [comment]
As far as the Intel storage driver, Lets focus on things we at System76 have control over first. If it does have something to do with our service we can move the issue to that repo and start getting things ready for the engineers.

**Comment by Dekamir** [comment]
**Disabling system76-power and rebooting stops any harddisk rebooting.

But, Pop still enables APM at every boot.** It doesn't even recognises that it did, though. The Disks app doesn't recognise it and reports as its off.
I can reboot to Windows and disable it, but it re-enables it at boot.

I just did a fresh install. I had to force the APM to the maximum level on the Disks app to even be able to install Pop to my harddrive. It keeps the harddrive restarts from 1 second to 5-ish seconds.

_TL;DR: system76-power is responsible for power cycles. The APM value is still re-enabled on every boot, which it shouldn't. (may be from Intel Storage driver)._

**Comment by bflanagin** [comment]
Well that's a step in the right direction at least. APM is always a mixed bag in Linux.  I'm hesitant to remove any modules that are loaded automatically for your hardware. 

A few things you can try. (sense we don't have the hardware to test on)

1. Use hdparm to tweak the APM settings similar to what you're doing in Windows. [manpage](https://man7.org/linux/man-pages/man8/hdparm.8.html), [stackexchange article](https://unix.stackexchange.com/questions/410264/hdd-idle-settings-using-hdparm-apm-suspend)
2. Go into the systems bios and modify the APM settings there. This will ensure the settings will survive reboots and work in both Linux and Windows. 

If you go the hdparm route it would be good to know what settings you ended up using as it will be useful if others have similar problems.

**Comment by bflanagin** [comment]
@Dekamir, we have a possible fix for the system76-power issues you've run into. However, because we don't have the specific hardware you're using we can't confirm the fix without your help.  If you're willing follow the steps below and report back with your findings. 

```
git clone https://github.com/pop-os/pop
cd pop
./scripts/apt add apm
sudo apt upgrade

```
If all goes well it will install the fix, I'm not sure if it will re-enable the service so be sure to run `sudo systemctl enable system76-power' before rebooting. Once the machine reboots you shouldn't see the APM errors on the drives.

If you do, could you supply us with the logs that detail the issue (assuming its being logged of course.)

**Comment by Dekamir** [comment]
It updated the system and I enabled system76-power. My harddrives don't spin down anymore.

hdparm reports APM_level as 254, before and after enabling system76-power, which means don't spin down and work at full power. 255 apparently means no APM. I will be doing more tests.

I don't know what to log, since it's just lagging because disks stop. If there's a general logging on Pop that I can find, I don't know where it is.

The script caused some shell components to pop up on my Applications view. They all open Desktop settings.
![Screenshot from 2021-07-15 17-16-57](https://user-images.githubusercontent.com/28929300/125803347-8293fb4c-15b3-4a37-b229-1c9c66b8a4a5.png)

**Comment by bflanagin** [comment]
Ah, well those new desktop files aren't a part of the system76-power changes, they are a part of the work being done to make Cosmic more accessible. 

As far as the hard drive issues I agree with your assessment, the APM settings must be set by something before system76-power and is connected to a global default. As far as the changes made to the service I'll approve the change and get the PR merged since it did solve some of the issue.

Shortly after the merge you'll get an error message when you try to update to avoid this run the following:

```
cd <to the pop directory you cloned>
./scripts/apt remove apm
sudo apt upgrade

```
This will downgrade your system's system76-power service so you will probably need to disable system76-power again until the update its fully released. 

` sudo systemctl disable system76-power` 

As far as your hardware is concerned you can edit the /etc/hdparm.conf file to make your settings persistent. The file has a good amount of documentation so setting things the way that works best for your situation shouldn't be an issue.

**Comment by bflanagin** [comment]
Note on logging: Most logs are stored in /var/log/ , but there are some helpful commands like 'dmesg' and 'journalctl' which parse the logs and can give you good debugging info. The journalctl command can be very verbose so you may want to run 'journalctl -h' before trying to dig through the output.
